### PR TITLE
Auction Module Unit Tests

### PR DESCRIPTION
**Pull request by philipjames44**
Closes: #172 

## Description

This PR introduces unit tests for the auction module.

### CODE REVIEW

File: x/auction/types/auction.go [hunk 1]
```diff
@@ -84,33 +76,25 @@ func (b *Bid) ValidateBasic() error {
...
 	if !strings.HasPrefix(b.SaleTokenMinimumAmount.Denom, gravitytypes.GravityDenomPrefix) {
-		return sdkerrors.Wrapf(ErrInvalidTokenBeingBidOn, "sale token: %s", b.SaleTokenMinimumAmount)
+		return sdkerrors.Wrapf(ErrInvalidTokenBeingBidOn, "sale token: %s", b.SaleTokenMinimumAmount.String())
 	}
 
 	if !b.SaleTokenMinimumAmount.IsPositive() {
 		return sdkerrors.Wrapf(ErrMinimumAmountMustBePositive, "sale token amount: %s", b.SaleTokenMinimumAmount.String())
 	}
 
-	if b.TotalFulfilledSaleTokens.Amount.IsNegative() {
```
Comment (EricBolten) [inline x/auction/types/auction.go:102]: Let's leave the negative checks in to make me feel safer.

**Review by EricBolten — COMMENTED** [review]
A few comments and suggestions.

**Comment by EricBolten** [comment]
Quick note: update go.mod conflicts and add unit testing to workflow.

File: x/auction/keeper/keeper_test.go [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+				priceDecreaseBlockInterval: uint64(10),
+				fundingModuleAccount:       permissionedFunder.GetName(),
+				proceedsModuleAccount:      permissionedReciever.GetName(),
+			},
+			expectedError: sdkerrors.Wrapf(auctionTypes.ErrCannotStartTwoAuctionsForSameDenomSimultaneously, "Denom: %s", auctionedSaleTokens.Denom),
+			utilityFunctions: func() {
+				// Mock initial bank keeper fund transfer
+				suite.mockSendCoinsFromModuleToModule(ctx, permissionedFunder.GetName(), auctionTypes.ModuleName, sdk.NewCoins(auctionedSaleTokens))
+
+				// Start auction
```
[the remark this answers is not in the document]
Comment (philipjames44) [reply x/auction/keeper/keeper_test.go:396]: As I started writing this out with utility functions it became a bit harder to intuitively read through the code and understand what was happening at each step, but if we feel abstracting this logic away is best I don't feel super strongly one way or the other so happy to change it.

File: x/auction/keeper/msg_server_test.go [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+
+	// Start auction
+	decreaseRate := sdk.MustNewDecFromStr("0.05")
+	blockDecreaseInterval := uint64(5)
+	err := auctionKeeper.BeginAuction(ctx, auctionedSaleTokens, decreaseRate, blockDecreaseInterval, permissionedFunder.GetName(), permissionedReciever.GetName())
+	require.Nil(err)
+
+	// Submit a bid
+	auctionID := uint32(1)
+	bidder := "cosmos16zrkzad482haunrn25ywvwy6fclh3vh7r0hcny"
```
[the remark this answers is not in the document]
Comment (philipjames44) [reply x/auction/keeper/msg_server_test.go]: Ended up generalizing the scope of these cosmos addresses for the tests so they're not repeated as magic strings everywhere, however setting up a cosmos keyring and generating keys with cosmos specific prefixes seemed like a bit of overkill; the code flow also becomes ever so slightly non deterministic with stuff like mnemonic generation which seems not particularly ideal for unit tests. 

Similar to the above comment though, don't feel super strongly about constant test addresses vs auto generated ones so happy to update if we prefer the latter.

File: x/auction/keeper/msg_server.go [hunk 4]
```diff
@@ -17,10 +16,9 @@ var _ types.MsgServer = Keeper{}
 func (k Keeper) SubmitBid(c context.Context, msg *types.MsgSubmitBidRequest) (*types.MsgSubmitBidResponse, error) {
 	ctx := sdk.UnwrapSDKContext(c)
 
-	// Verify signer is the same as the bidder (this validates both the bidder and signer addresses)
-	signer := msg.MustGetSigner()
-	if !signer.Equals(sdk.AccAddress(msg.Bidder)) {
-		return &types.MsgSubmitBidResponse{}, sdkerrors.Wrapf(types.ErrSignerDifferentFromBidder, "Signer: %s, Bidder: %s", msg.GetSigner(), msg.GetBidder())
+	bidder, err := sdk.AccAddressFromBech32(msg.Bidder)
```
Comment (EricBolten) [inline x/auction/keeper/msg_server.go:19]: If we're dropping `MustGetSigner` and using the bidder for `GetSigners`, we should retrieve it using `MustGetBidder` here. Although I would consider keeping it as `MustGetSigner` and just using the bidder as the signer field.

File: .github/workflows/unit-test.yml [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        os: [ubuntu-latest]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/setup-go@v3
+        with:
+          go-version: ${{ matrix.go-version }}
+      - uses: actions/checkout@v3
+      - run: go test github.com/peggyjv/sommelier/.../x/...
+
+  test-cache:
```
Comment (EricBolten) [inline .github/workflows/unit-test.yml:25]: What's the reasoning behind there being two different unit testing jobs (`test` and `test-cache`)? Also is this based off of the workflow that used to exist in this repo for unit testing or is this from elsewhere?

**Review by EricBolten — COMMENTED** [review]
Two small comments.

File: .github/workflows/unit-test.yml (hunk 5 above)
Comment (philipjames44) [reply .github/workflows/unit-test.yml:25]: Ah yeah I was trying to cache tests. Ill remove this one. And no this is a new workflow for unit testing I spun up after looking at how a couple other repos were doing golang unit tests. If theres another template we prefer to use I can update to that

File: x/auction/keeper/msg_server.go (hunk 4 above)
Comment (philipjames44) [reply x/auction/keeper/msg_server.go:19]: sure that's fair, will update.

Comment (philipjames44) [reply x/auction/keeper/msg_server.go]: Updated.

File: .github/workflows/unit-test.yml (hunk 5 above)
Comment (philipjames44) [reply .github/workflows/unit-test.yml]: Updated.

**Review by EricBolten — APPROVED** [review]
LGTM

**Pull request merged**